File: Language/C#/Book/tmp/LINQ实战/LinqInAction/Chapter13.CSharp - LinqBooks/LinqBooks.Web/Controls/AuthorList.ascx.cs
#region Imports

using System;
using System.Data;
using System.Configuration;
using System.Collections;
using System.Collections.Generic;
using System.Linq;
using System.Web;
using System.Web.Security;
using System.Web.UI;
using System.Web.UI.WebControls;
using System.Web.UI.WebControls.WebParts;
using System.Web.UI.HtmlControls;

using LinqBooks.Entities;

#endregion Imports

namespace LinqBooks.Web.Controls
{
  public partial class AuthorList : System.Web.UI.UserControl
  {
    public class AuthorEventArgs : EventArgs
    {
      public Guid AuthorId { get; set; }
    }

    public event EventHandler<AuthorEventArgs> DeleteAuthor;

    private IEnumerable<Author> _Authors;

    public IEnumerable<Author> Authors
    {
      set { _Authors = value; }
    }

    protected void Page_Load(object sender, EventArgs e)
    {

    }

    public override void DataBind()
    {
      base.DataBind();
      if (_Authors != null)
      {
        DisplayAuthors();
      }
    }

    private void DisplayAuthors()
    {
      if (_Authors != null)
      {
        GridViewAuthors.DataSource =
          _Authors.Select(author => new { ID = author.ID, FullName = author.FullName });
      }
      else
      {
        GridViewAuthors.DataSource = null;
      }
      GridViewAuthors.DataBind();
    }

    protected void GridViewAuthors_RowDeleting(object sender, GridViewDeleteEventArgs e)
    {
      OnDeleteAuthor((Guid)GridViewAuthors.DataKeys[e.RowIndex].Value);
    }

    private void OnDeleteAuthor(Guid authorId)
    {
      if (DeleteAuthor != null)
      {
        DeleteAuthor(this, new AuthorEventArgs { AuthorId = authorId });
      }
    }
  }
}

File: Language/C#/Book/tmp/LINQ实战/Chapter13.VB - LinqBooks/LinqBooks.Web/Controls/AuthorList.ascx.vb
Imports LinqBooks.Entities

Partial Public Class AuthorList
  Inherits System.Web.UI.UserControl

  Public Class AuthorEventArgs
    Inherits EventArgs

    Private _AuthorID As Guid
    Public Property AuthorID() As Guid
      Get
        Return _AuthorID
      End Get
      Set(ByVal value As Guid)
        _AuthorID = value
      End Set
    End Property
  End Class

  Public Event DeleteAuthor As EventHandler(Of AuthorEventArgs)

  Private _Authors As IEnumerable(Of Author)

  Public WriteOnly Property Authors() As IEnumerable(Of Author)
    Set(ByVal value As IEnumerable(Of Author))
      _Authors = value
    End Set
  End Property

  Protected Sub Page_Load(ByVal sender As Object, ByVal e As System.EventArgs) Handles Me.Load

  End Sub

  Public Overrides Sub DataBind()
    MyBase.DataBind()
    If Not _Authors Is Nothing Then
      DisplayAuthors()
    End If
  End Sub

  Private Sub DisplayAuthors()
    If Not _Authors Is Nothing Then
      GridViewAuthors.DataSource = _
        _Authors.Select(Function(author) New With {.ID = author.ID, .FullName = author.FullName})
    Else
      GridViewAuthors.DataSource = Nothing
    End If
    GridViewAuthors.DataBind()
  End Sub

  Protected Sub GridViewAuthors_RowDeleting(ByVal sender As Object, ByVal e As GridViewDeleteEventArgs)
    OnDeleteAuthor(CType(GridViewAuthors.DataKeys(e.RowIndex).Value, Guid))
  End Sub

  Private Sub OnDeleteAuthor(ByVal authorId As Guid)
    RaiseEvent DeleteAuthor(Me, New AuthorEventArgs With {.AuthorID = authorId})
  End Sub
End Class